scvi/dataset/dataset.py

```diff
@@ -53,10 +57,18 @@ def download_and_preprocess(self):
     def collate_fn(self, batch):
         indexes = np.array(batch)
         X = torch.FloatTensor(self.X[indexes]) if self.dense else torch.FloatTensor(self.X[indexes].toarray())
-        return X, torch.FloatTensor(self.local_means[indexes]), \
-            torch.FloatTensor(self.local_vars[indexes]), \
-            torch.LongTensor(self.batch_indices[indexes]), \
-            torch.LongTensor(self.labels[indexes])
+        if not self.got_spatial:
+            return X, torch.FloatTensor(self.local_means[indexes]), \
+                   torch.FloatTensor(self.local_vars[indexes]), \
+                   torch.LongTensor(self.batch_indices[indexes]), \
+                   torch.LongTensor(self.labels[indexes])
```

Comment:
Better to return dummy values for x_coord and y_coord, like we do for batch_indices when there aren't batches, rather than branching

Reply:
@Edouard360  Could we move `collate_fn` to the `Inference` classes? (and override it in `VariationalInferenceFish` with this one?)

For that matter, we move all the `DataLoaders` logic (e.g. `TrainTestDataLoader`) to the `Inference` class? It seems like almost every Inference class e.g. (`VariationalInference`) needs a different data loader any way -- it'd be nice not to have two separate hierarchies of classes.

Reply:
`TrainTestDataLoader` is also used in `ClassifierInference`, which serves to: 1) the "alternate" training scheme `SemiSupervisedModel` in which at the end of each epoch, we train only the classifier weights on top of the latent space 2) if we wanted to do a NN classifier on top of a VAE or on raw data.

I see what you mean. It is possible to move the functionalities of `DataLoaders`  into `Inference`, only keeping the `DataLoadersWrapper` class.

`collate_fn` is a  method of pytorch's `Dataset` class used by pytorch's dataloaders however... I am not sure to understand how to make it part of Inference.